File: OSCommon/org/optimizationservices/oscommon/representationparser/OSqLWriter.cs
using System;
using System.Xml;
using System.Collections;

using org.optimizationservices.oscommon.util;
using org.optimizationservices.oscommon.localinterface;
using org.optimizationservices.oscommon.datastructure.osquery;
using org.optimizationservices.oscommon.datastructure.osentity;

namespace org.optimizationservices.oscommon.representationparser{
	/// <summary>
	/// The <c>OSqLWriter</c> class is used to construct an instance that follows 
	/// the OSqL format. 
	/// @author Jun Ma	
	/// @version 1.0, 09/01/2005
	/// @since OS 1.0
	/// @copyright (c) 2005
	/// </summary>
	public class OSqLWriter : OSgLWriter{
		/// <summary>
		/// m_eOSqL holds the OSqL root element.
		/// </summary>
		protected internal XmlElement m_eOSqL = null;

        /// <summary>
		/// constructor. 
		/// </summary>
		public OSqLWriter(){
			m_document = XMLUtil.createNewDocument();
			if(OSParameter.XSLT_LOCATION != null && OSParameter.XSLT_LOCATION.Length > 0){
				XmlProcessingInstruction xsltPI = m_document.CreateProcessingInstruction("xml-stylesheet",
					"type=\"text/xsl\" href = \"" + OSParameter.XSLT_LOCATION + "OSqL.xslt\"");
				m_document.AppendChild(xsltPI);     
			}
			m_eOSqL = createOSqLRoot();
			m_document.AppendChild(m_eOSqL);
		}//constructor
		
		/// <summary>
		/// set the OSQuery, a standard os query interface. 
		/// </summary>
		/// <param name="osQuery">holds the standard os query interface. </param>
		/// <returns>whether the OSQuery is set successfully. </returns>
		public bool setOSQuery(OSQuery osQuery){
			if(osQuery.standard != null){
				if(!setStandardQuery(osQuery.getStandardQuery())){ 
					throw new Exception("setStandardQuery Unsuccessful");
				}
			}
			else if(osQuery.xQuery != null){
				if(!setXQuery(osQuery.getXQuery())){ 
					throw new Exception("setXQuery Unsuccessful");
				}
			}
			else{
				throw new Exception("No OSQuery specified");
			}
			return true;
		}//setOSQuery

		/// <summary>
		/// set the standard query. 
		/// @see org.optimizationservices.oscommon.datastructure.osquery.StandardQuery
		/// </summary>
		/// <param name="standardQuery">holds the standard query in the StandardQuery data structure. </param>
		/// <returns>whether the standard query is set successfully. </returns>
		public bool setStandardQuery(StandardQuery standardQuery){
			try{
				if(standardQuery != null){
					XMLUtil.removeAllChildren(m_eOSqL);
					XmlElement eStandard = m_document.CreateElement("standard");
					if(standardQuery.currentState != null && standardQuery.currentState.Length > 0){				
						XmlElement eCurrentState = m_document.CreateElement("currentState");
						eCurrentState.AppendChild(m_document.CreateTextNode(standardQuery.currentState));
						eStandard.AppendChild(eCurrentState);					
					}
					if(standardQuery.optimization != null){
						XmlElement eOptimization = m_document.CreateElement("optimization");
					
						if(standardQuery.optimization.description != null && standardQuery.optimization.description.Length > 0){
							XmlElement eDescription = m_document.CreateElement("description");
							eDescription.AppendChild(m_document.CreateTextNode(standardQuery.optimization.description));
							eOptimization.AppendChild(eDescription);
						}
						if(standardQuery.optimization.variableType != null && standardQuery.optimization.variableType.Length > 0){
							XmlElement eVariableType = m_document.CreateElement("variableType");
							eVariableType.AppendChild(m_document.CreateTextNode(standardQuery.optimization.variableType));
							eOptimization.AppendChild(eVariableType);
						}
						if(standardQuery.optimization.objectiveType != null && standardQuery.optimization.objectiveType.Length > 0){
							XmlElement eObjectiveType = m_document.CreateElement("objectiveType");
							eObjectiveType.AppendChild(m_document.CreateTextNode(standardQuery.optimization.objectiveType));
							eOptimization.AppendChild(eObjectiveType);
						}
						if(standardQuery.optimization.constraintType != null && standardQuery.optimization.constraintType.Length > 0){
							XmlElement eConstraintType = m_document.CreateElement("constraintType");
							eConstraintType.AppendChild(m_document.CreateTextNode(standardQuery.optimization.constraintType));
							eOptimization.AppendChild(eConstraintType);
						}
						if(standardQuery.optimization.objectiveLinearity != null && standardQuery.optimization.objectiveLinearity.Length > 0){
							XmlElement eObjectiveLinearity = m_document.CreateElement("objectiveLinearity");
							eObjectiveLinearity.AppendChild(m_document.CreateTextNode(standardQuery.optimization.objectiveLinearity));
							eOptimization.AppendChild(eObjectiveLinearity);
						}
						if(standardQuery.optimization.constraintLinearity != null && standardQuery.optimization.constraintLinearity.Length > 0){
							XmlElement eConstraintLinearity = m_document.CreateElement("constraintLinearity");
							eConstraintLinearity.AppendChild(m_document.CreateTextNode(standardQuery.optimization.constraintLinearity));
							eOptimization.AppendChild(eConstraintLinearity);
						}
						if(standardQuery.optimization.objectiveDifferentiability != null && standardQuery.optimization.objectiveDifferentiability.Length > 0){
							XmlElement eObjectiveDifferentiability = m_document.CreateElement("objectiveDifferentiability");
							eObjectiveDifferentiability.AppendChild(m_document.CreateTextNode(standardQuery.optimization.objectiveDifferentiability));
							eOptimization.AppendChild(eObjectiveDifferentiability);
						}
						if(standardQuery.optimization.constraintDifferentiability != null && standardQuery.optimization.constraintDifferentiability.Length > 0){
							XmlElement eConstraintDifferentiability = m_document.CreateElement("constraintDifferentiability");
							eConstraintDifferentiability.AppendChild(m_document.CreateTextNode(standardQuery.optimization.constraintDifferentiability));
							eOptimization.AppendChild(eConstraintDifferentiability);
						}
						if(standardQuery.optimization.parameterType != null && standardQuery.optimization.parameterType.Length > 0){
							XmlElement eParameterType = m_document.CreateElement("parameterType");
							eParameterType.AppendChild(m_document.CreateTextNode(standardQuery.optimization.parameterType));
							eOptimization.AppendChild(eParameterType);
						}
						if(standardQuery.optimization.stochasticity != null && standardQuery.optimization.stochasticity.Length > 0){
							XmlElement eStochasticity = m_document.CreateElement("stochasticity");
							eStochasticity.AppendChild(m_document.CreateTextNode(standardQuery.optimization.stochasticity));
							eOptimization.AppendChild(eStochasticity);
						}
						if(standardQuery.optimization.realTime == true){
							XmlElement eRealTime = m_document.CreateElement("realTime");
							eRealTime.AppendChild(m_document.CreateTextNode("true"));
							eOptimization.AppendChild(eRealTime);
						}
						if(standardQuery.optimization.functionType != null){
							XmlElement eFunctionType = m_document.CreateElement("functionType");
						
							if(standardQuery.optimization.functionType.general == true){
								XmlElement eGeneral = m_document.CreateElement("general");
								eGeneral.AppendChild(m_document.CreateTextNode("true"));
								eFunctionType.AppendChild(eGeneral);
							}
							if(standardQuery.optimization.functionType.basic == false){
								XmlElement eBasic = m_document.CreateElement("basic");
								eBasic.AppendChild(m_document.CreateTextNode("false"));
								eFunctionType.AppendChild(eBasic);
							}
							if(standardQuery.optimization.functionType.arithmetic == false){
								XmlElement eArithmetic = m_document.CreateElement("arithmetic");
								eArithmetic.AppendChild(m_document.CreateTextNode("false"));
								eFunctionType.AppendChild(eArithmetic);
							}
							if(standardQuery.optimization.functionType.trigonometric == true){
								XmlElement eTrigonometric = m_document.CreateElement("trigonometric");
								eTrigonometric.AppendChild(m_document.CreateTextNode("true"));
								eFunctionType.AppendChild(eTrigonometric);
							}
							if(standardQuery.optimization.functionType.statistics == true){
								XmlElement eStatistics = m_document.CreateElement("statistics");
								eStatistics.AppendChild(m_document.CreateTextNode("true"));
								eFunctionType.AppendChild(eStatistics);
							}
							if(standardQuery.optimization.functionType.probability == true){
								XmlElement eProbability = m_document.CreateElement("probability");
								eProbability.AppendChild(m_document.CreateTextNode("true"));
								eFunctionType.AppendChild(eProbability);
							}
							if(standardQuery.optimization.functionType.relationalAndLogic == true){
								XmlElement eRelationalAndLogic = m_document.CreateElement("relationalAndLogic");
								eRelationalAndLogic.AppendChild(m_document.CreateTextNode("true"));
								eFunctionType.AppendChild(eRelationalAndLogic);
							}
							if(standardQuery.optimization.functionType.userFunction == true){
								XmlElement eUserFunction = m_document.CreateElement("userFunction");
								eUserFunction.AppendChild(m_document.CreateTextNode("true"));
								eFunctionType.AppendChild(eUserFunction);
							}
							if(standardQuery.optimization.functionType.simulation == true){
								XmlElement eSimulation = m_document.CreateElement("simulation");
								eSimulation.AppendChild(m_document.CreateTextNode("true"));
								eFunctionType.AppendChild(eSimulation);
							}
						
							eOptimization.AppendChild(eFunctionType);
						}
						if(standardQuery.optimization.specialStructure != null){
							XmlElement eSpecialStructure = m_document.CreateElement("specialStructure");
						
							if(standardQuery.optimization.specialStructure.semidefiniteProgram == true){
								XmlElement eSemidefiniteProgram = m_document.CreateElement("semidefiniteProgram");
								eSemidefiniteProgram.AppendChild(m_document.CreateTextNode("true"));
								eSpecialStructure.AppendChild(eSemidefiniteProgram);
							}
							if(standardQuery.optimization.specialStructure.coneProgram == true){
								XmlElement eConeProgram = m_document.CreateElement("coneProgram");
								eConeProgram.AppendChild(m_document.CreateTextNode("true"));
								eSpecialStructure.AppendChild(eConeProgram);
							}						
							if(standardQuery.optimization.specialStructure.disjunctiveProgram == true){
								XmlElement eDisjunctiveProgram = m_document.CreateElement("disjunctiveProgram");
								eDisjunctiveProgram.AppendChild(m_document.CreateTextNode("true"));
								eSpecialStructure.AppendChild(eDisjunctiveProgram);
							}
						
							eOptimization.AppendChild(eSpecialStructure);
						}
						if(standardQuery.optimization.specialAlgorithm != null){
							XmlElement eSpecialAlgorithm = m_document.CreateElement("specialAlgorithm");
						
							if(standardQuery.optimization.specialAlgorithm.decomposition == true){
								XmlElement eDecomposition = m_document.CreateElement("decomposition");
								eDecomposition.AppendChild(m_document.CreateTextNode("true"));
								eSpecialAlgorithm.AppendChild(eDecomposition);
							}
							if(standardQuery.optimization.specialAlgorithm.globalOptimization == true){
								XmlElement eGlobalOptimization = m_document.CreateElement("globalOptimization");
								eGlobalOptimization.AppendChild(m_document.CreateTextNode("true"));
								eSpecialAlgorithm.AppendChild(eGlobalOptimization);
							}
							if(standardQuery.optimization.specialAlgorithm.dynamicProgramming == true){
								XmlElement eDynamicProgramming = m_document.CreateElement("dynamicProgramming");
								eDynamicProgramming.AppendChild(m_document.CreateTextNode("true"));
								eSpecialAlgorithm.AppendChild(eDynamicProgramming);
							}
							if(standardQuery.optimization.specialAlgorithm.heuristic == true){
								XmlElement eHeuristic = m_document.CreateElement("heuristic");
								eHeuristic.AppendChild(m_document.CreateTextNode("true"));
								eSpecialAlgorithm.AppendChild(eHeuristic);
							}						
						
							eOptimization.AppendChild(eSpecialAlgorithm);
						}
						eStandard.AppendChild(eOptimization);
					}
					m_eOSqL.AppendChild(eStandard);
				}
				else{
					XMLUtil.removeChildrenByName(m_eOSqL, "standard");
				}
			}
			catch(Exception e){
				Console.WriteLine(e.ToString());
				return false;
			}
			return true;
		}//setStandardQuery

		/// <summary>
		/// Set the XQuery. 
		/// </summary>
		/// <param name="xQuery">holds the XQuery in a string. </param>
		/// <returns>whether the XQuery is set successfully or not. </returns>
		public bool setXQuery(string xQuery){
			try{
				if(xQuery != null && xQuery.Length > 0){
					XMLUtil.removeAllChildren(m_eOSqL);
					XmlElement eXQuery = m_document.CreateElement("xQuery");
					eXQuery.AppendChild(m_document.CreateTextNode(xQuery));
					m_eOSqL.AppendChild(eXQuery);
			
				}
				else{
					XMLUtil.removeChildrenByName(m_eOSqL, "xQuery");
				}
			}
			catch(Exception e){
				Console.WriteLine(e.ToString());
				return false;
			}
			return true;
		}//setXQuery
	
		

		/// <summary>
		/// Create the OSqL root element and its most basic required structure.
		/// </summary>
		/// <returns>the OSqL element.</returns>
		protected XmlElement createOSqLRoot(){
			XmlElement eOSqL = XMLUtil.createOSxLRootElement(m_document, "osql");
			return eOSqL;
		}//createOSqLRoot

	}//class OSqLWriter
}//namespace

File: OSCommon/src/org/optimizationservices/oscommon/representationparser/OSqLWriter.java
/**
 * @(#)OSbLWriter 1.0 03/14/2004
 *
 * Copyright (c) 2004
 */
package org.optimizationservices.oscommon.representationparser;
import org.optimizationservices.oscommon.datastructure.osentity.FunctionType;
import org.optimizationservices.oscommon.datastructure.osentity.OptimizationInfo;
import org.optimizationservices.oscommon.datastructure.osentity.SpecialAlgorithm;
import org.optimizationservices.oscommon.datastructure.osentity.SpecialStructure;
import org.optimizationservices.oscommon.datastructure.osquery.StandardQuery;
import org.optimizationservices.oscommon.localinterface.OSQuery;
import org.optimizationservices.oscommon.util.OSParameter;
import org.optimizationservices.oscommon.util.XMLUtil;
import org.w3c.dom.Element;
import org.w3c.dom.ProcessingInstruction;
/**
 * The <code>OSqLWriter</code> class is used to construct an instance that follows 
 * the OSqL format. 
 * </p>
 *
 * @author Robert Fourer, Jun Ma, Kipp Martin
 * @version 1.0, 03/14/2004
 * @since OS 1.0
 */
public class OSqLWriter extends OSgLWriter{

	/**
	 * m_eOSqL holds the OSqL root element.
	 */
	protected Element m_eOSqL = null;

	
	/**
	 * constructor. 
	 *
	 */
	public OSqLWriter(){
		m_document = XMLUtil.createNewDocument();
		if(OSParameter.XSLT_LOCATION != null && OSParameter.XSLT_LOCATION.length() > 0){
			ProcessingInstruction xsltPI = m_document.createProcessingInstruction("xml-stylesheet",
	           "type=\"text/xsl\" href = \"" + OSParameter.XSLT_LOCATION + "OSqL.xslt\"");
			m_document.appendChild(xsltPI);     
		}
		m_eOSqL = createOSqLRoot();
		m_document.appendChild(m_eOSqL);
	}//constructor
	
	/**
	 * set the OSQuery, a standard os query interface. 
	 * @param osQuery holds the standard os query interface. 
	 * @return whether the OSQuery is set successfully. 
	 */
	public boolean setOSQuery(OSQuery osQuery) throws Exception{
		if(osQuery.standard != null){
			if(!setStandardQuery(osQuery.getStandardQuery())){ 
	   			throw new Exception("setStandardQuery Unsuccessful");
			}
		}
		else if(osQuery.xQuery != null){
			if(!setXQuery(osQuery.getXQuery())){ 
	   			throw new Exception("setXQuery Unsuccessful");
			}
		}
		else{
			throw new Exception("No OSQuery specified");
		}
		return true;
	}//setOSQuery
		
	/**
	 * set the standard query. 
	 * @param standardQuery holds the standard query in the StandardQuery data structure. 
	 * @return whether the standard query is set successfully. 
	 * @see org.optimizationservices.oscommon.datastructure.osquery.StandardQuery
	 */
	public boolean setStandardQuery(StandardQuery standardQuery){
		try{
			if(standardQuery != null){
				XMLUtil.removeAllChildren(m_eOSqL);
				Element eStandard = m_document.createElement("standard");
				if(standardQuery.currentState != null && standardQuery.currentState.length() > 0){				
					Element eCurrentState = m_document.createElement("currentState");
					eCurrentState.appendChild(m_document.createTextNode(standardQuery.currentState));
					eStandard.appendChild(eCurrentState);					
				}
				if(standardQuery.optimization != null){
					Element eOptimization = m_document.createElement("optimization");
					
					if(standardQuery.optimization.description != null && standardQuery.optimization.description.length() > 0){
						Element eDescription = m_document.createElement("description");
						eDescription.appendChild(m_document.createTextNode(standardQuery.optimization.description));
						eOptimization.appendChild(eDescription);
					}
					if(standardQuery.optimization.variableType != null && standardQuery.optimization.variableType.length() > 0){
						Element eVariableType = m_document.createElement("variableType");
						eVariableType.appendChild(m_document.createTextNode(standardQuery.optimization.variableType));
						eOptimization.appendChild(eVariableType);
					}
					if(standardQuery.optimization.objectiveType != null && standardQuery.optimization.objectiveType.length() > 0){
						Element eObjectiveType = m_document.createElement("objectiveType");
						eObjectiveType.appendChild(m_document.createTextNode(standardQuery.optimization.objectiveType));
						eOptimization.appendChild(eObjectiveType);
					}
					if(standardQuery.optimization.constraintType != null && standardQuery.optimization.constraintType.length() > 0){
						Element eConstraintType = m_document.createElement("constraintType");
						eConstraintType.appendChild(m_document.createTextNode(standardQuery.optimization.constraintType));
						eOptimization.appendChild(eConstraintType);
					}
					if(standardQuery.optimization.objectiveLinearity != null && standardQuery.optimization.objectiveLinearity.length() > 0){
						Element eObjectiveLinearity = m_document.createElement("objectiveLinearity");
						eObjectiveLinearity.appendChild(m_document.createTextNode(standardQuery.optimization.objectiveLinearity));
						eOptimization.appendChild(eObjectiveLinearity);
					}
					if(standardQuery.optimization.constraintLinearity != null && standardQuery.optimization.constraintLinearity.length() > 0){
						Element eConstraintLinearity = m_document.createElement("constraintLinearity");
						eConstraintLinearity.appendChild(m_document.createTextNode(standardQuery.optimization.constraintLinearity));
						eOptimization.appendChild(eConstraintLinearity);
					}
					if(standardQuery.optimization.objectiveDifferentiability != null && standardQuery.optimization.objectiveDifferentiability.length() > 0){
						Element eObjectiveDifferentiability = m_document.createElement("objectiveDifferentiability");
						eObjectiveDifferentiability.appendChild(m_document.createTextNode(standardQuery.optimization.objectiveDifferentiability));
						eOptimization.appendChild(eObjectiveDifferentiability);
					}
					if(standardQuery.optimization.constraintDifferentiability != null && standardQuery.optimization.constraintDifferentiability.length() > 0){
						Element eConstraintDifferentiability = m_document.createElement("constraintDifferentiability");
						eConstraintDifferentiability.appendChild(m_document.createTextNode(standardQuery.optimization.constraintDifferentiability));
						eOptimization.appendChild(eConstraintDifferentiability);
					}
					if(standardQuery.optimization.parameterType != null && standardQuery.optimization.parameterType.length() > 0){
						Element eParameterType = m_document.createElement("parameterType");
						eParameterType.appendChild(m_document.createTextNode(standardQuery.optimization.parameterType));
						eOptimization.appendChild(eParameterType);
					}
					if(standardQuery.optimization.stochasticity != null && standardQuery.optimization.stochasticity.length() > 0){
						Element eStochasticity = m_document.createElement("stochasticity");
						eStochasticity.appendChild(m_document.createTextNode(standardQuery.optimization.stochasticity));
						eOptimization.appendChild(eStochasticity);
					}
					if(standardQuery.optimization.realTime == true){
						Element eRealTime = m_document.createElement("realTime");
						eRealTime.appendChild(m_document.createTextNode("true"));
						eOptimization.appendChild(eRealTime);
					}
					if(standardQuery.optimization.functionType != null){
						Element eFunctionType = m_document.createElement("functionType");
						
						if(standardQuery.optimization.functionType.general == true){
							Element eGeneral = m_document.createElement("general");
							eGeneral.appendChild(m_document.createTextNode("true"));
							eFunctionType.appendChild(eGeneral);
						}
						if(standardQuery.optimization.functionType.basic == false){
							Element eBasic = m_document.createElement("basic");
							eBasic.appendChild(m_document.createTextNode("false"));
							eFunctionType.appendChild(eBasic);
						}
						if(standardQuery.optimization.functionType.arithmetic == false){
							Element eArithmetic = m_document.createElement("arithmetic");
							eArithmetic.appendChild(m_document.createTextNode("false"));
							eFunctionType.appendChild(eArithmetic);
						}
						if(standardQuery.optimization.functionType.trigonometric == true){
							Element eTrigonometric = m_document.createElement("trigonometric");
							eTrigonometric.appendChild(m_document.createTextNode("true"));
							eFunctionType.appendChild(eTrigonometric);
						}
						if(standardQuery.optimization.functionType.statistics == true){
							Element eStatistics = m_document.createElement("statistics");
							eStatistics.appendChild(m_document.createTextNode("true"));
							eFunctionType.appendChild(eStatistics);
						}
						if(standardQuery.optimization.functionType.probability == true){
							Element eProbability = m_document.createElement("probability");
							eProbability.appendChild(m_document.createTextNode("true"));
							eFunctionType.appendChild(eProbability);
						}
						if(standardQuery.optimization.functionType.relationalAndLogic == true){
							Element eRelationalAndLogic = m_document.createElement("relationalAndLogic");
							eRelationalAndLogic.appendChild(m_document.createTextNode("true"));
							eFunctionType.appendChild(eRelationalAndLogic);
						}
						if(standardQuery.optimization.functionType.userFunction == true){
							Element eUserFunction = m_document.createElement("userFunction");
							eUserFunction.appendChild(m_document.createTextNode("true"));
							eFunctionType.appendChild(eUserFunction);
						}
						if(standardQuery.optimization.functionType.simulation == true){
							Element eSimulation = m_document.createElement("simulation");
							eSimulation.appendChild(m_document.createTextNode("true"));
							eFunctionType.appendChild(eSimulation);
						}
						
						eOptimization.appendChild(eFunctionType);
					}
					if(standardQuery.optimization.specialStructure != null){
						Element eSpecialStructure = m_document.createElement("specialStructure");
						
						if(standardQuery.optimization.specialStructure.semidefiniteProgram == true){
							Element eSemidefiniteProgram = m_document.createElement("semidefiniteProgram");
							eSemidefiniteProgram.appendChild(m_document.createTextNode("true"));
							eSpecialStructure.appendChild(eSemidefiniteProgram);
						}
						if(standardQuery.optimization.specialStructure.coneProgram == true){
							Element eConeProgram = m_document.createElement("coneProgram");
							eConeProgram.appendChild(m_document.createTextNode("true"));
							eSpecialStructure.appendChild(eConeProgram);
						}						
						if(standardQuery.optimization.specialStructure.disjunctiveProgram == true){
							Element eDisjunctiveProgram = m_document.createElement("disjunctiveProgram");
							eDisjunctiveProgram.appendChild(m_document.createTextNode("true"));
							eSpecialStructure.appendChild(eDisjunctiveProgram);
						}
						
						eOptimization.appendChild(eSpecialStructure);
					}
					if(standardQuery.optimization.specialAlgorithm != null){
						Element eSpecialAlgorithm = m_document.createElement("specialAlgorithm");
						
						if(standardQuery.optimization.specialAlgorithm.decomposition == true){
							Element eDecomposition = m_document.createElement("decomposition");
							eDecomposition.appendChild(m_document.createTextNode("true"));
							eSpecialAlgorithm.appendChild(eDecomposition);
						}
						if(standardQuery.optimization.specialAlgorithm.globalOptimization == true){
							Element eGlobalOptimization = m_document.createElement("globalOptimization");
							eGlobalOptimization.appendChild(m_document.createTextNode("true"));
							eSpecialAlgorithm.appendChild(eGlobalOptimization);
						}
						if(standardQuery.optimization.specialAlgorithm.dynamicProgramming == true){
							Element eDynamicProgramming = m_document.createElement("dynamicProgramming");
							eDynamicProgramming.appendChild(m_document.createTextNode("true"));
							eSpecialAlgorithm.appendChild(eDynamicProgramming);
						}
						if(standardQuery.optimization.specialAlgorithm.heuristic == true){
							Element eHeuristic = m_document.createElement("heuristic");
							eHeuristic.appendChild(m_document.createTextNode("true"));
							eSpecialAlgorithm.appendChild(eHeuristic);
						}						
						
						eOptimization.appendChild(eSpecialAlgorithm);
					}
					eStandard.appendChild(eOptimization);
				}
				m_eOSqL.appendChild(eStandard);
			}
			else{
				XMLUtil.removeChildrenByName(m_eOSqL, "standard");
			}
		}
		catch(Exception e){
			e.printStackTrace();
			return false;
		}
		return true;
	}//setStandardQuery
	
	
 	/**
	 * Set the XQuery. 
	 * 
	 * 
	 * @param xQuery holds the XQuery in a string. 
	 * @return whether the XQuery is set successfully or not. 
	 */
	public boolean setXQuery(String xQuery){
		try{
			if(xQuery != null && xQuery.length() > 0){
				XMLUtil.removeAllChildren(m_eOSqL);
				Element eXQuery = m_document.createElement("xQuery");
				eXQuery.appendChild(m_document.createTextNode(xQuery));
				m_eOSqL.appendChild(eXQuery);
			}
			else{
				XMLUtil.removeChildrenByName(m_eOSqL, "xQuery");
			}
		}
		catch(Exception e){
			e.printStackTrace();
			return false;
		}
		return true;
	}//setXQuery

	
	/**
	 * Create the OSqL root element and its most basic required structure.
	 * 
	 * @return the OSqL element.
	 */
	protected Element createOSqLRoot(){
		Element eOSqL = XMLUtil.createOSxLRootElement(m_document, "osql");
		return eOSqL;
	}//createOSqLRoot
	
	/**
	 * main for test purposes.
	 *
	 * </p>
	 *
	 * @param argv command line arguments.
	 */
	public static void main(String[] argv){
		OSqLWriter osqlWriter = new OSqLWriter();
		String sFileName = OSParameter.CODE_HOME + "OSRepository/test/osql/osql.osql";

		if(!osqlWriter.setXQuery("xQuery xQuery xQuery")) System.out.println("setXQuery!");
		StandardQuery q = new StandardQuery();
		q.currentState = "busy";
		OptimizationInfo opt = new OptimizationInfo();
		opt.constraintType = "generalRange";
		opt.variableType = "mixedInteger";
		opt.objectiveLinearity = "linear";
		opt.functionType = new FunctionType();
		opt.functionType.probability = true;
		opt.specialAlgorithm = new SpecialAlgorithm();
		opt.specialAlgorithm.decomposition = true;
		opt.specialStructure = new SpecialStructure();
		opt.specialStructure.disjunctiveProgram = true;
		q.optimization = opt;
		if(!osqlWriter.setStandardQuery(q)) System.out.println("setStandardQuery!");
	
		////////////////////

		OSqLReader osqlReader = new OSqLReader(false);
		if(!osqlReader.readFile(sFileName)) System.out.println("reading");
		OSQuery osQuery;
		try {
			osQuery = osqlReader.getOSQuery();
			if(!osqlWriter.setOSQuery(osQuery)) System.out.println("set OSQuery");
		} 
		catch (Exception e) {
			e.printStackTrace();
		}
		////////////////////
		
		//osqlWriter.writeToStandardOutput();
		System.out.println(osqlWriter.writeToString());
	
	}//main
	
}//class OSqLWriter
